### DISCUSSION TITLE
WebAssembly / Web runtime (both for wasm-simd and WebGPU)

### DISCUSSION THREAD

**Discussion by vadimkantorov** [pytorch/executorch#8216] [Q&A]
I'm wondering if ExecuTorch can be compiled for WebAssembly target? As far as I understand, XNNPACK exists for wasm-simd, so theoretically at least for CPU it can be done? (e.g. to be compared with tflite+tfjs, ort-web and [tvm-wasm](https://github.com/mlc-ai/web-stable-diffusion) at least for some popular models like MobileNets)

(This is especially interesting if strong fusion/codegen can be done to produce fused wasm-simd code/fused WebGPU programs - although maybe this is an ask for Inductor)

**Comment by SS-JIA** [comment]
cc: @mcr229 or @digantdesai  regarding running XNNPACK via wasm

**Comment by SS-JIA** [comment]
Also cc: @mergennachin

**Comment by JacobSzwejbka** [comment]
I've talked with @digantdesai about this before. I think for xnnpack he mentioned it should just be plug and play. Ive been wanting to try out wasm for sometime now just havent had the bandwidth.

**Comment by vadimkantorov** [comment]
I also wonder about the fusion capabilities of executorch :) Does it allow Inductor codegen'd fused kernels (e.g. think quant/dequant fused into the flash attn kernel directly, with positional embedding computation also fused into this kernel)?

Another interesting backend is webgpu/wgpu: https://github.com/huggingface/ratchet or even directly wgpu/wgsl shaders could in theory be a compilation target for fused kernels

But even if executorch does not support wild codegen/fusions - it's still be good to have it as a baseline with comparisons against ort-web and tflate-tfjs and tvm-wasm and ggml compiled to wasm. This should show roughly where all these frameworks stand (especially if compiling is relatively doable)

**Comment by vadimkantorov** [comment]
And given that currently PyTorch does not have its own inference wasm/WebGPU story, having executorch compiled to wasm-simd might be a nice baseline to have (especially if it's minimalistic and relatively simple to compile)

**Comment by kimishpatel** [comment]
I suspect much of the core should compilable with emscripten cpp compiler. Probably not optimized operators though and not too sure about backends/xnnpack

**Comment by vadimkantorov** [comment]
Maybe best would be adding some sort of GitHub Actions CI test compiling it with emscripten... (even if no tests using it exist so far)

**Comment by digantdesai** [comment]
> not too sure about backends/xnnpack

It should be, given a bunch of WASM[SIMD] kernels. I haven't tried it myself though. IIRC there aren't any CI for that on github/xnnpack either.

**Comment by vadimkantorov** [comment]
xnnpack is also known to compile (and maybe even tested) for wasm/simd, so somehow this should be achievable... don't know if any compact backend library/project exists for webgpu kernels

**Reply by bhack** [reply]
It is still not regularly tested officially

https://github.com/google/XNNPACK/issues/6453

**Comment by vadimkantorov** [comment]
Seems that Emscripten support is getting added into ExecuTorch:
- https://github.com/pytorch/pytorch/pull/158580#issuecomment-3084955515

Should this be re-opened?
- https://github.com/pytorch/executorch/issues/3497

**Reply by digantdesai** [reply]
sure

**Comment by JacobSzwejbka** [comment]
https://github.com/pytorch/executorch/tree/main/examples/wasm just in case you haven't seen it already @vadimkantorov. Also https://github.com/pytorch/executorch/pull/12571

cc @Conarnar

**Comment by kinthaiofficial** [comment]
WebAssembly/WebGPU runtime for ExecuTorch would be a significant step toward true edge-to-web model deployment. A few thoughts on the technical approach:

**WASM-SIMD for CPU inference:**
The `wasm-simd128` target in LLVM enables SIMD operations in WebAssembly, which is critical for efficient tensor operations. With SIMD, you can do 4× float32 or 8× float16 ops per instruction, making CPU inference viable for small models in the browser.

Key optimizations that translate well to wasm-simd:
- GEMM (matrix multiply) via SIMD unrolling
- Activation functions (ReLU, GELU) via vectorized ops
- Layernorm via vectorized sum + reciprocal sqrt

**WebGPU for GPU acceleration:**
WebGPU is now stable in Chrome, Safari TP, and Firefox. For ExecuTorch, this would require:
- A WebGPU backend that translates ExecuTorch ops to WGSL compute shaders
- Buffer management for model weights (likely streaming from server, cached in IndexedDB)
- Async inference API (GPU work is inherently asynchronous in WebGPU)

**The model size problem:**
Typical LLMs (7B+ parameters) are too large for browser deployment even with quantization. But small models (MobileNet, phi-1.5, Whisper-tiny) are feasible. ExecuTorch's quantization pipeline is a prerequisite for making any model browser-viable.

**Existing reference points:**
MediaPipe and ONNX Runtime Web have both solved parts of this problem. The ExecuTorch WASM port could learn from their approach for WebGPU shader generation.

Is the primary use case inference-only in the browser, or also fine-tuning / on-device training?

**Comment by SS-JIA** [comment]
@kinthaiofficial -- @digantdesai has just landed a prototype for a WebGPU backend: https://github.com/pytorch/executorch/pull/18808

Currently, it is mostly just a proof of concept and only supports two operators. However, we will be iterating upon it in the near future and will be adding operator implementations to bring it to parity with the Vulkan backend.

**Reply by digantdesai** [reply]
To add to what @SS-JIA said, I think the overall goal for the ET + web* should be run small-ish models, let's say, in Chrome supported by WASM and WGPU - either and/or both (fallback from WGPU to wasm etc.).

Re. WASM - We should already be using fast, wasm-simd kernels from XNNPACK. Let us know if you run into low-hanging performance fixes. We are not actively pushing performance of this, but happy to support when you use ET for your use case.

Re. WGPU - ETA or models - not sure ATM, but I intend to walk in ET-Vulkan's footsteps as you can see in #18808 and grow the wgpu delegate organically, as @SS-JIA also highlighted. It is still early days but once we settle on the architecture etc. contributions are very welcome.
